src/java.base/share/classes/java/lang/runtime/Carrier.java

```diff
@@ -129,15 +135,324 @@
+        } else {
+            index = carrierShape.intOffset() + componentCounts.intCount();
+        }
+
+        MethodHandle component = components[index];
+
+        if (filter != null) {
+            component = MethodHandles.filterReturnValue(component, filter);
+        }
+
+        component = MethodHandles.explicitCastArguments(component,
+                methodType(ptype, Object.class));
+
+        return component;
+    }
+
+    /**
+     * Factory for carriers that are backed by int[] and Object[]. This strategy is
+     * used when the number of components exceeds {@link Carrier#MAX_OBJECT_COMPONENTS}.
      */
     private static class CarrierArrayFactory {
+        /**
+         * Lookup used to define and reference the carrier array methods.
+         */
+        private static final MethodHandles.Lookup LOOKUP;
+
+        /**
+         * Unsafe access.
+         */
+        private static final Unsafe UNSAFE;
+
+        static {
+            LOOKUP = MethodHandles.lookup();
+            UNSAFE = Unsafe.getUnsafe();
+
+            try {
+                CONSTRUCTOR = LOOKUP.findConstructor(CarrierArray.class,
+                        methodType(void.class, int.class, int.class));
+                GET_LONG = LOOKUP.findVirtual(CarrierArray.class, "getLong",
+                        methodType(long.class, int.class));
+                PUT_LONG = LOOKUP.findVirtual(CarrierArray.class, "putLong",
+                        methodType(CarrierArray.class, int.class, long.class));
+                GET_INTEGER = LOOKUP.findVirtual(CarrierArray.class, "getInteger",
+                        methodType(int.class, int.class));
+                PUT_INTEGER = LOOKUP.findVirtual(CarrierArray.class, "putInteger",
+                        methodType(CarrierArray.class, int.class, int.class));
+                GET_OBJECT = LOOKUP.findVirtual(CarrierArray.class, "getObject",
+                        methodType(Object.class, int.class));
+                PUT_OBJECT = LOOKUP.findVirtual(CarrierArray.class, "putObject",
+                        methodType(CarrierArray.class, int.class, Object.class));
+            } catch (ReflectiveOperationException ex) {
+                throw new AssertionError("carrier static init fail", ex);
+            }
+        }
+
+        /*
+         * Constructor accessor MethodHandles.
+         */
+        private static final MethodHandle CONSTRUCTOR;
+        private static final MethodHandle GET_LONG;
+        private static final MethodHandle PUT_LONG;
+        private static final MethodHandle GET_INTEGER;
+        private static final MethodHandle PUT_INTEGER;
+        private static final MethodHandle GET_OBJECT;
+        private static final MethodHandle PUT_OBJECT;
+
+        /**
+         * Wrapper object for carrier arrays.
+         */
+        private static class CarrierArray {
+            /**
+             * Carrier for longs and integers.
+             */
+            private final int[] integers;
```

Comment:
I believe (@rose00 correct me if I'm wrong) that an `int[]` doesn't guarantee proper 64-but alignment of the first element in the array in 32-bit platforms. We ran into similar issues with the MemorySegment API, I believe. I think using a `long[]` if you want explicit 64-bit alignment is the safest choice.

Reply:
Makes sense. Easy change.

Reply:
Done